### PR TITLE
storage: switch API from ledger.Path to flow.RegisterID

### PR DESCRIPTION
**Pull request by SaveTheRbtz**
This PR removes key hashing into the `ledger.Path` and brings `storage` API closer to `storage2`'s `GetPayload` and `BatchSetPayload`.

While here, also:
* removed the non-batched `SavePayload` method.
* removed mirroring of inputs to outputs from `GetRegisterValuesResponse`.

ref: https://github.com/onflow/flow-archive/issues/99

### CODE REVIEW

File: models/convert/registers.go [hunk 1]
```diff
@@ -0,0 +1,64 @@
+// Copyright 2023 Dapper Labs
```
Comment (SaveTheRbtz) [inline models/convert/registers.go:1]: TODO: working on the unit tests for this.

File: api/protobuf/api.proto [hunk 2]
```diff
@@ -88,16 +88,18 @@ message GetEventsResponse {
...
-  ];
+  repeated bytes registers = 3 [(validate.rules).repeated.min_items = 1];
 }
 
 message GetRegisterValuesResponse {
+  reserved 2;
+  reserved "paths";
+
   uint64 height = 1;
-  repeated bytes paths = 2;
```
Comment (SaveTheRbtz) [inline api/protobuf/api.proto:100]: There is no reason to mirror back inputs

File: service/storage/operations.go [hunk 3]
```diff
@@ -187,18 +180,17 @@ func (l *Library) RetrieveEvents(height uint64, types []flow.EventType, events *
...
-		key := EncodeKey(PrefixPayload, path, height)
+		key := EncodeKey(PrefixPayload, register, height)
 		it := tx.NewIterator(badger.IteratorOptions{
 			PrefetchSize:   0,
 			PrefetchValues: false,
 			Reverse:        true,
 			AllVersions:    false,
 			InternalAccess: false,
-			Prefix:         key[:1+pathfinder.PathByteSize],
+			Prefix:         key[:len(key)-8],
```
Comment (zhangchiqing) [inline service/storage/operations.go:193]: magic 8?

File: service/storage/operations_internal_test.go [hunk 4]
```diff
@@ -670,10 +677,8 @@ func TestLibrary_SaveAndRetrievePayload(t *testing.T) {
 			codec: codec,
 		}
 
-		unknownPath := ledger.Path{0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff, 0xff}
-
-		var got ledger.Payload
-		err = db.View(l.RetrievePayload(mocks.GenericHeight, unknownPath, &got))
+		var got flow.RegisterValue
+		err = db.View(l.RetrievePayload(mocks.GenericHeight, mocks.GenericRegister(10), &got))
```
Comment (pattyshack) [inline service/storage/operations_internal_test.go:681]: maybe define a variable `unknownId := 10` to make it clear this should error

File: api/protobuf/api.proto [hunk 5]
```diff
@@ -88,16 +88,18 @@ message GetEventsResponse {
...
-    (validate.rules).repeated.min_items = 1,
-    (validate.rules).repeated.items.bytes.len = 32
-  ];
+  repeated bytes registers = 3 [(validate.rules).repeated.min_items = 1];
 }
 
 message GetRegisterValuesResponse {
-  uint64 height = 1;
-  repeated bytes paths = 2;
+  reserved 1, 2;
```
Comment (pattyshack) [inline api/protobuf/api.proto:99]: fancy new pb definition syntax.  this is new-ish (to me)

File: api/protobuf/api.proto [hunk 6]
```diff
@@ -88,16 +88,18 @@ message GetEventsResponse {
...
+  reserved "paths";
```
Comment (pattyshack) [inline api/protobuf/api.proto:100]: reserving the number is good enough?

File: models/convert/registers.go (hunk 1 above)
Comment (pattyshack) [inline models/convert/registers.go:1]: we can skip the per file license if the license is at the repo root

File: service/storage/operations.go (hunk 3 above)
Comment (pattyshack) [reply service/storage/operations.go:193]: +1

File: service/index/reader.go [hunk 7]
```diff
@@ -110,24 +109,25 @@ func (r *Reader) Values(height uint64, paths []ledger.Path) ([]ledger.Value, err
...
 	g.SetLimit(ConcurrentPathReadLimit)
-	for i, path := range paths {
-		i, path := i, path
+	for i, reg := range regs {
+		i, reg := i, reg
 		g.Go(func() error {
-			err = r.db.View(func(tx *badger.Txn) error {
-				var payload ledger.Payload
-				err := r.lib.RetrievePayload(height, path, &payload)(tx)
+			err := r.db.View(func(tx *badger.Txn) error {
```
Comment (pattyshack) [inline service/index/reader.go:119]: just return the err directly?

File: testing/mocks/generic.go [hunk 8]
```diff
@@ -201,32 +200,67 @@ func GenericLedgerPath(index int) ledger.Path {
...
 
-	var values []ledger.Value
+	registers := make(flow.RegisterIDs, 0, number)
 	for i := 0; i < number; i++ {
-		value := make(ledger.Value, 32)
-		binary.BigEndian.PutUint64(value[0:], random.Uint64())
-		binary.BigEndian.PutUint64(value[8:], random.Uint64())
-		binary.BigEndian.PutUint64(value[16:], random.Uint64())
-		binary.BigEndian.PutUint64(value[24:], random.Uint64())
+		registers = append(registers, flow.RegisterID{
```
Comment (pattyshack) [inline testing/mocks/generic.go:226]: this should call GenericRegister, and `flow.RegisterID{...}` should be in GenericRegister's body.

File: testing/mocks/generic.go [hunk 9]
```diff
@@ -201,32 +200,67 @@ func GenericLedgerPath(index int) ledger.Path {
...
+	return GenericRegisters(index + 1)[index]
+}
+
+func GenericRegisterValues(number int) []flow.RegisterValue {
+	// Ensure consistent deterministic results.
+	random := rand.New(rand.NewSource(3))
+
+	values := make([]flow.RegisterValue, 0, number)
+	for i := 0; i < number; i++ {
+		values = append(values, []byte(fmt.Sprintf("value-%d", random.Uint64())))
```
Comment (pattyshack) [inline testing/mocks/generic.go:245]: same issue as GenericRegister

**Review by pattyshack — APPROVED** [review]

File: api/protobuf/api.proto (hunk 6 above)
Comment (SaveTheRbtz) [reply api/protobuf/api.proto:100]: it's just a due diligence-- there are no downsides for doing this

File: testing/mocks/generic.go (hunk 8 above)
Comment (SaveTheRbtz) [reply testing/mocks/generic.go]: Oh, you are breaking my hear^W^W this tests' quadratic, cubic, quatric, etc, complexity Patrick!

File: service/storage/encoding.go [hunk 10]
```diff
@@ -25,6 +25,12 @@ func EncodeKey(prefix uint8, segments ...interface{}) []byte {
 		case flow.StateCommitment:
 			val = make([]byte, 32)
 			copy(val, s[:])
+		case flow.RegisterID:
+			val = make([]byte, 0, len(s.Owner)+1+len(s.Key)+1)
+			val = append(val, s.Owner...)
+			val = append(val, '/')
+			val = append(val, s.Key...)
+			val = append(val, '/')
```
Comment (zhangchiqing) [inline service/storage/encoding.go:33]: can we combine the code with [the lookupKey here](https://github.com/onflow/flow-archive/pull/104/files#diff-ec5ed77af8eef26b4b6e9d31f8ee434f8d53cf639d0bd9f7cc779e8fe7401435R24-R27)

File: service/storage/operations.go [hunk 11]
```diff
@@ -173,18 +170,17 @@ func (l *Library) RetrieveEvents(height uint64, types []flow.EventType, events *
...
-		key := EncodeKey(PrefixPayload, path, height)
+		key := EncodeKey(PrefixPayload, register, height)
 		it := tx.NewIterator(badger.IteratorOptions{
 			PrefetchSize:   0,
 			PrefetchValues: false,
 			Reverse:        true,
 			AllVersions:    false,
 			InternalAccess: false,
-			Prefix:         key[:1+pathfinder.PathByteSize],
+			Prefix:         key[:len(key)-heightBinarySize],
```
Comment (zhangchiqing) [inline service/storage/operations.go:183]:
```suggestion
			// In order to find the latest payload up to the given height, we need to iterate
			// through all the keys for the registers on different height.
			// so the prefix is the encoded register ID
			Prefix:         key[:len(key)-heightBinarySize],
```

Comment (zhangchiqing) [reply service/storage/operations.go:183]: Here we first encode the register with the height, and later we subtract the bytes for the height. 

Would it be more clear if we just encode the register ID?

```
  Prefix: EncodeRegisterID(register)
```

File: service/storage/encoding.go (hunk 10 above)
Comment (SaveTheRbtz) [reply service/storage/encoding.go:33]: This code will go away after we switch payloads to the storage2 API.  storage depending on APIs from storage2 is a leaking abstraction it only makes sense if we factor this code out into a separate package, but that package would not make sense when we remove storage1 payloads API later

tl;dr is that case statement is implementation detail and would go away.

**Review by zhangchiqing — APPROVED** [review]
